### DISCUSSION TITLE
Getting "random" `Dependency Lookup Warnings` for composer deps from private Satis

### DISCUSSION THREAD

**Discussion by simPod** [renovatebot/renovate#23034] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Docker tag `:35`, ~I even tried going down to `:35.128`~

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

Renovate started having issues with fetching updates from our private Satis. But randomly. Sometimes it succeeds and sometimes it does not.

The all runners connectivity to our Satis (https://composer.domain.eu/) works smoothly.

So the update PRs from Renovate look like this now:

<img width="621" alt="image" src="https://github.com/renovatebot/renovate/assets/327717/e2ae1b5a-200c-41e6-afdb-9cee347ea95c">

Where each update is adding or removing info about dep lookup failure.

<img width="1624" alt="image" src="https://github.com/renovatebot/renovate/assets/327717/472aea5c-ce07-4778-b861-ac2327313875">

When it succeeds I can see the correct req to `https://composer.domain.eu/p2/private/package.json` in `DEBUG: http statistics` log section. When it fails, the url is missing.

### Relevant debug logs

<details><summary>Logs</summary>

This part is not in the log when it succeeds. Otherwise, it tries to fetch it from packagist.org where the package is not published. Maybe it even skips our Satis?

```
DEBUG: GET https://packagist.org/p2/private/package.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=228) (repository=a/b/c)
DEBUG: Datasource 404 (repository=a/b/c)
       "datasource": "packagist",
       "packageName": "private/package",
       "url": "https://packagist.org/p2/private/package.json"
DEBUG: Failed to look up packagist package private/package (repository=a/b/c, packageFile=composer.json, dependency=private/package)
DEBUG: GET https://packagist.org/p2/private/package~dev.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=380) (repository=a/b/c)
```

```
               {
                 "currentValue": "^0.1.2",
                 "currentVersion": "0.1.2",
                 "datasource": "packagist",
                 "depName": "private/package",
                 "depType": "require",
                 "fixedVersion": "0.1.2",
                 "lockedVersion": "0.1.2",
                 "packageName": "private/package",
                 "registryUrl": "https://composer.domain.eu",
                 "registryUrls": [
                   "https://composer.domain.eu",
                   "https://packagist.org"
                 ],
                 "versioning": "composer",
                 "warnings": [
                   {
                     "topic": "private/package",
                     "message": "Failed to look up packagist package private/package"
                   }
                 ],
                 "updates": []
               },
```

```
DEBUG: Found package lookup warnings (repository=a/b/c)
       "warnings": [
         "Failed to look up packagist package private/package",
         "Failed to look up packagist package private/package2",
         "Failed to look up packagist package private/package3",
         "Failed to look up packagist package private/package4"
       ],
       "warningFiles": ["composer.json"]
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
Those warnings don't necessarily apply to _that_ update. In fact if you have a PR then it implies it was looked up fine. It's letting you know that _some_ dependency you have has lookup problems. For example it might be a Docker Hub image and you occasionally get rate limited. So yes it looks like it's due to packagist lookups. For this particular package you want it only looked up from your private registry, not packagist.org?

**Reply by simPod** [reply]
Yes, there's no reason it's looked up from Packagist.org. But I supposed the Renovate first tries the one specified in `"registryUrl": "https://composer.domain.eu",`

And what is weird is that it succeeds once and fails on some following run.

When the run is successful, it does not try to hit packagist.org (can see that from `http statistics`).  
When the run fails, it hits packagist.org and not composer.domain.eu (at least composer.domain.eu is missing from http-statistics, there's nothing like `(GET,4xx)` or sth).

**Reply by rarkins** [reply]
That's pretty strange. Have you set up any Redis package cache? If not then the cache will be local (disk). It's possible that it's trying from cache - there's also some logged stats about cache hits and misses. Try comparing the stats from two runs.

The "packageFiles with updates" log is also useful, especially for the failed case (look for `skipReason` field).

BTW if the lookup completely failed then I'd expect it to autoclose the PRs.

**Reply by simPod** [reply]
> BTW if the lookup completely failed then I'd expect it to autoclose the PRs.

Yup it does that. Other PRs that are irrelevant to the private package are kept open and only desc changes as shown here https://user-images.githubusercontent.com/327717/249741987-472aea5c-ce07-4778-b861-ac2327313875.png

I have local cache. Will check that.

**Reply by simPod** [reply]
I tried fully purging the cache but the result is still the same 🤔 .

There's no `skipReason` for `private/package`.

```json
               {
                 "currentValue": "^0.1.2",
                 "currentVersion": "0.1.2",
                 "datasource": "packagist",
                 "depName": "private/package",
                 "depType": "require",
                 "fixedVersion": "0.1.2",
                 "lockedVersion": "0.1.2",
                 "packageName": "private/package",
                 "registryUrl": "https://composer.domain.eu",
                 "registryUrls": [
                   "https://composer.domain.eu",
                   "https://packagist.org"
                 ],
                 "versioning": "composer",
                 "warnings": [
                   {
                     "topic": "private/package",
                     "message": "Failed to look up packagist package private/package"
                   }
                 ],
                 "updates": []
               }
```

**Reply by simPod** [reply]
I tried 35.100 and it's running ok 😕

**Reply by simPod** [reply]
35.110 as well

**Reply by simPod** [reply]
35.129 is ok. 35.140 is not

**Reply by simPod** [reply]
36 is also problematic. Trying to isolate 35.129<x<35.140

**Reply by simPod** [reply]
The last working is 35.133

I suspect https://github.com/renovatebot/renovate/pull/22877

**Comment by amenk** [comment]
As you say this is happen randomly: How often is the satis update running? How often renovate? Is there eventually a point in time where the available packages files is not available? Maybe try monitoring this with repeated curl (or something similar) downloads of the file?

**Reply by simPod** [reply]
Now as I investigated it is not random. It is rather each second run. The renovate runs every 10 minutes (`*/10 * * * *`)

<img width="303" alt="image" src="https://github.com/renovatebot/renovate/assets/327717/7a94e4e4-dc74-4ffd-964e-e973cf5a7e0b">

The connection to satis is stable (also, there are no issues in previous renovate versions. Like timeouts etc.). 

Not sure what you mean by satis update. If you mean refreshing satis database from vcs, it's instant. Reactive to vcs changes. However, there are mostly no changes happening now.
